code/modules/xenoarchaeology/xenoartifact.dm

```diff
@@ -0,0 +1,461 @@
+			if(get_dist(get_turf(src), get_turf(M)) > max_range)   
+				true_target -= M
+		charge+=charge_mod
+		for(var/datum/xenoartifact_trait/t as() in traits)//Minor & malfunction traits aren't apart of the target loop
+			if(!istype(t, /datum/xenoartifact_trait/major))
+				t.activate(src, user, user)
+				if(logging)
+					log_game("[src] activated [istype(t, /datum/xenoartifact_trait/malfunction) ? "malfunction" : "minor"] trait [t] at [world.time]. Located at [x] [y] [z]")
+		charge = (charge+charge_req)/1.9 //Not quite an average. Generally produces better results.     
+		for(var/atom/M in true_target) //target loop
+			create_beam(M) //Indicator beam, points to target, M
+			for(var/datum/xenoartifact_trait/major/t as() in traits) //Major traits
+				if(logging)
+					log_game("[src] activated major trait [t] at [world.time]. Located at [x] [y] [z]")
+				t.activate(src, M, user)
+			if(!(get_trait(/datum/xenoartifact_trait/minor/aura))) //Quick fix for bug that selects multiple targets for noraisin
+				break
+		COOLDOWN_START(src, xenoa_cooldown, cooldown+cooldownmod)
+	charge = 0
+	true_target = list()
+
+/obj/item/xenoartifact/proc/generate_traits(var/list/blacklist_traits, malf = FALSE) //Generate traits outside of blacklist. Malf = TRUE if you want malfunctions
+	var/datum/xenoartifact_trait/new_trait //Selection
+	var/datum/xenoartifact_trait/holder //holder for blacklisting solution
+	
+	var/list/activators = subtypesof(/datum/xenoartifact_trait/activator)
+	activators -= blacklist_traits
+	if(activators.len < 1)
+		log_game("An almost impossible event has occured. [src] has failed to generate any traits!")
+		return
+	new_trait = pick(activators)
+	blacklist_traits += new_trait //Add chosen trait to blacklist
+	traits += new new_trait
+	holder = new new_trait //Add traits blacklist to our blacklist
+	blacklist_traits += holder.blacklist_traits
+	qdel(holder) //Discard trait and prep for next
+	special_desc = initial(new_trait.desc) ? "[special_desc] [initial(new_trait.desc)]" : special_desc
+
+	var/minor_desc
+	var/list/minors = subtypesof(/datum/xenoartifact_trait/minor)
+	for(var/X in 1 to 3) //Minors
+		minors -= blacklist_traits
+		if(minors.len < 1)
+			log_game("An almost impossible event has occured. [src] has failed to generate any traits!")
+			return
+		new_trait = pick(minors)
+		blacklist_traits += new_trait
+		traits += new new_trait
+		holder = new new_trait
+		blacklist_traits += holder.blacklist_traits
+		qdel(holder)
+		if(!minor_desc && initial(new_trait.desc))
+			minor_desc = initial(new_trait.desc)
+		if(!touch_desc) //This is weird but doing it other ways causes issues regarding early exit.
+			var/datum/xenoartifact_trait/D = new new_trait
+			if(D.on_touch(src, src))
+				touch_desc = new new_trait
+			else
+				qdel(D)
+
+	special_desc = minor_desc ? "[special_desc] [minor_desc] material." : "[special_desc] material"
+
+	var/list/majors = subtypesof(/datum/xenoartifact_trait/major)
```

Comment:
I'm not gonna make a comment on every use of typeof and subtypeof but, again, hold them in static list.